plugins/crd/crd_utils.go

```diff
@@ -46,31 +55,83 @@ func newCustomResourceDefinition(plugin *Plugin, FullName, Group, Version, Plura
 		GetOpenAPIDefinitions: v1.GetOpenAPIDefinitions,
 	})
 
-	crdClient := plugin.apiclientset.ApiextensionsV1beta1().CustomResourceDefinitions()
+	crdClient := plugin.apiclientset
+	// Add NSM CRD version
+	crd.ObjectMeta.Annotations = map[string]string{
+		nsmCRDVersionAnnotationKey: nsmCRDVersion,
+	}
+	// Starting with 1.11.X SpecReplicasPath and StatusReplicasPath have become mandatory fields,
+	// but since NewCustomResourceDefinition function does not set them, CRD creation fails in +1.11.0 k8s clusters.
+	// As a workaround, setting these two fields manually here.
+	crd.Spec.Subresources.Scale.SpecReplicasPath = ".spec.replicas"
+	crd.Spec.Subresources.Scale.StatusReplicasPath = ".status.replicas"
+	if err := createCRDObject(crd, crdClient); err != nil {
+		plugin.Log.Errorf("fail to create CRD with error: %#v", err)
+		return err
+	}
+
+	return nil
+}
 
+func createCRDObject(newCRD *apiextv1beta1.CustomResourceDefinition, crdClient *apiextcs.Clientset) error {
 	// First check if the CRD already exists
```

Comment:
There's a bit more going on here in createCRDObject() and updateCRD(). @edwarnicke had mentioned the possibility of race conditions during my previous patch. I wasn't as concerned then, but I wonder now that we're adding version capability into these and potentially updating the definitions we should add a lock in createCRDObject() to ensure multiple NSMs don't race with each other. What do you think @sbezverk?

Reply:
@mestery we cannot lock CRD object since it is responsibility of API server to sort this out. API server implements optimistic concurrency and eventual consistency models which will deal with racing in k8s cluster. But you brought up a case which is not handled by this PR, it is when two NSM daemonsets start on a new cluster and tries to create CRDs. One will succeed but the second will fail since the object already exists and due to race between getting and creating there is a short but still time.  It is not big deal as second NSM will crash and then when container gets restarted it will come up. Also I can address this situation adding a check for error type. If error AlreadyExist returned from create, then it was a race and no action is required. WDYT?

Reply:
I like your solution around AlreadyExist, that works for me.